oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core;
+
+
+import java.net.URL;
+import java.util.List;
+
+/**
+ * A base {@link ClaimAccessor} for the "claims" the Authorization Server can make about
+ * its configuration, used either in OpenID Connect Discovery 1.0 or OAuth 2.0 Authorization
+ * Server Metadata.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see ClaimAccessor
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc8414#section-2">2. Authorization Server Metadata</a>
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata">3. OpenID Provider Metadata</a>
+ */
+public interface AuthorizationServerMetadataClaimAccessor extends ClaimAccessor {
```

Comment:
I feel like we should merge this into `OAuth2AuthorizationServerMetadataClaimAccessor`. Then we can remove `AuthorizationServerMetadataClaimAccessor`, `AuthorizationServerMetadataClaimNames` and `AbstractAuthorizationServerMetadata`. This change will simplify the hierarchy.

Although `revocation_endpoint`, `revocation_endpoint_auth_methods_supported` and `code_challenge_methods_supported` are not defined in OIDC Provider Metadata, I believe an OIDC provider will support these configurations either way.

What are your thoughts on this proposed change?

Reply:
👍  definitely agree with the proposed change

Quick research:
- Okta supports those three claims in their OIDC configuration
- [Google](https://accounts.google.com/.well-known/openid-configuration) supports `revocation_endpoint` and `code_challenge_methods_supported`
- [Microsoft](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) supports none of them

Reply:
One caveat, removing `AbstractAuthorizationServerMetadata`, using `OAuth2AuthorizationServerConfiguration` instead in `OidcProviderConfiguration extends OAuth2AuthorizationServerConfiguration` 

Is only possible if we make `OAuth2AuthorizationServerConfiguration` not final. And I believe we want our model classes to be final?

Reply:
Our typical design process is to make things final (or restrict access) at the start. However, if there is need to open things up than we do only if it makes sense. This is one of those instances so it's ok to make `OAuth2AuthorizationServerConfiguration` non-final BUT make the constructor `protected`.

Reply:
So I've tried this, unfortunately the type hierarchy with generic builders does not allow to have this kind of hierarchy:

```
- OAuth2AuthorizationServerConfiguration
  - Builder (inner class)
  - static builder() method, returning a Builder
  - OidcProviderConfiguration extends OAuth2AuthorizationServerConfiguration
    - OidcBuilder (inner class, extends Builder) 
    - static builder() method, returning an OidcBuilder
```

Problems are:
- If `OidcBuilder` extends the concrete class `OAuth2AuthorizationServerConfiguration.Builder`, the builder pattern won't work as, for example, `builder.issuer("some string")` would always return the parent `Builder` type. I don't think we can get away without a base abstract builder, see [this Stackoverflow thread for reference](https://stackoverflow.com/questions/21086417/builder-pattern-and-inheritance) for reference.
- If both Builder and OidcBuilder inherit a common, generic, AbstractBuilder, then OidcProviderConfiguration.builder() return type becomes incompatible with that of `OAuth2AuthorizationServerConfiguration.builder()`.

If we really want to get rid of the `AbstractAuthorizationServerConfiguration`, we could have the abstract builder and default claims in the `AuthorizationServerMetadataClaimAccessor` interface.